common/src/main/java/io/netty/util/ResourceLeakDetector.java

```diff
@@ -451,4 +451,28 @@ static String newRecord(Object hint, int recordsToSkip) {
 
         return buf.toString();
     }
+
+    /**
+     * Calls {@link ResourceLeak#close()} and ensure the tracked object is not collected before and returns its return
+     * value. This is the proper way to close a {@link ResourceLeak} from within user code.
+     */
+    @SuppressWarnings("deprecation")
+    public static boolean close(ResourceLeak leak, Object trackedObject) {
+        // This should never be called with a null object, so just put in an assert to make it cheap to check.
+        assert trackedObject != null;
+
+        // This is kind of an ugly hack but the only way how to fix this without breaking the API for now.
+        //
+        // We need to actually do the null check of the trackedObject after we call leak.close() because otherwise
+        // we may get false-positives reported by the ResourceLeakDetector. This can happen as the JIT / GC may be able
+        // to figure out that we do not need the trackedObject anymore and so already enqueue it for collection before
+        // we actually get a chance to close the enclosing ResourceLeak.
+        //
+        // A more proper fix in the future is to change the ResourceLeak.close() method to take the tracked object as
+        // argument and so ensure it will not be collected before we had a chance to close the ResourceLeak.
+        //
+        // See also https://github.com/netty/netty/issues/6034
+        //
+        return leak.close() && trackedObject != null;
```

Comment:
Even better we could also add another method to `ResourceLeakDetector` which returns this new subclass and use it in our code. Then we mark `ResourceLeakDetector.open(...)` as deprecated. 

@Scottmitch let me do this and add as a separate commit to this pr so you can review more easily